File: frontend/src/api/metrics.ts
/**
 * API client for metrics endpoints.
 */

import type {
  MetricsResponse,
  TokenUsageResponse,
  ExecutionTimeResponse,
  CostAnalysisResponse,
  InsightsResponse,
  TrendsResponse,
  PerformanceResponse,
  ROIResponse,
  ProductivityResponse,
  HourlyMetricsResponse,
  DateRange,
  Granularity,
  GroupBy,
  PeriodComparison,
} from '../types/metrics';
import { API_CONFIG } from './config';

const BASE_URL = `${API_CONFIG.BASE_URL}/api/metrics`;

/**
 * Retorna métricas agregadas do projeto
 */
export async function getProjectMetrics(
  projectId: string,
  startDate?: string,
  endDate?: string
): Promise<MetricsResponse> {
  const params = new URLSearchParams();
  if (startDate) params.append('start_date', startDate);
  if (endDate) params.append('end_date', endDate);

  const url = `${BASE_URL}/project/${projectId}${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch project metrics: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Retorna uso de tokens agregado por período
 */
export async function getTokenUsage(
  projectId: string,
  period: DateRange = '7d',
  groupBy: 'hour' | 'day' | 'model' = 'day'
): Promise<TokenUsageResponse> {
  const params = new URLSearchParams({ period, group_by: groupBy });
  const url = `${BASE_URL}/tokens/${projectId}?${params.toString()}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch token usage: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Retorna tempos de execução por card/comando
 */
export async function getExecutionTimes(
  projectId: string,
  command?: string,
  limit: number = 100
): Promise<ExecutionTimeResponse> {
  const params = new URLSearchParams({ limit: limit.toString() });
  if (command) params.append('command', command);

  const url = `${BASE_URL}/execution-time/${projectId}?${params.toString()}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch execution times: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Retorna análise de custos detalhada
 */
export async function getCostAnalysis(
  projectId: string,
  groupBy: 'model' | 'command' | 'day' = 'model'
): Promise<CostAnalysisResponse> {
  const params = new URLSearchParams({ group_by: groupBy });
  const url = `${BASE_URL}/costs/${projectId}?${params.toString()}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch cost analysis: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Calcula tendências de uso de tokens
 */
export async function getTokenTrends(
  projectId: string,
  days: number = 7
): Promise<TrendsResponse> {
  const params = new URLSearchParams({ days: days.toString() });
  const url = `${BASE_URL}/trends/${projectId}?${params.toString()}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch token trends: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Analisa performance de execuções (percentis, outliers)
 */
export async function getExecutionPerformance(
  projectId: string,
  command?: string
): Promise<PerformanceResponse> {
  const params = new URLSearchParams();
  if (command) params.append('command', command);

  const url = `${BASE_URL}/performance/${projectId}${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch execution performance: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Calcula métricas de ROI (custo por card, eficiência, economia de tempo)
 */
export async function getROIMetrics(projectId: string): Promise<ROIResponse> {
  const url = `${BASE_URL}/roi/${projectId}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch ROI metrics: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Retorna métricas de produtividade (velocity, cycle time, throughput)
 */
export async function getProductivityMetrics(
  projectId: string,
  startDate?: string,
  endDate?: string
): Promise<ProductivityResponse> {
  const params = new URLSearchParams();
  if (startDate) params.append('start_date', startDate);
  if (endDate) params.append('end_date', endDate);

  const url = `${BASE_URL}/productivity/${projectId}${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch productivity metrics: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Gera insights automáticos sobre as métricas
 */
export async function getInsights(projectId: string): Promise<InsightsResponse> {
  const url = `${BASE_URL}/insights/${projectId}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch insights: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Retorna métricas agregadas por hora para um dia específico
 */
export async function getHourlyMetrics(
  projectId: string,
  targetDate?: string
): Promise<HourlyMetricsResponse> {
  const params = new URLSearchParams();
  if (targetDate) params.append('target_date', targetDate);

  const url = `${BASE_URL}/hourly/${projectId}${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch hourly metrics: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Preenche métricas retroativamente para execuções existentes
 */
export async function backfillMetrics(
  projectId: string,
  startDate?: string
): Promise<{ message: string; metricsCreated: number }> {
  const params = new URLSearchParams();
  if (startDate) params.append('start_date', startDate);

  const url = `${BASE_URL}/backfill/${projectId}${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await fetch(url, { method: 'POST' });

  if (!response.ok) {
    throw new Error(`Failed to backfill metrics: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Compara métricas entre dois períodos
 */
export async function comparePeriods(
  projectId: string,
  currentStart: string,
  currentEnd: string,
  previousStart: string,
  previousEnd: string
): Promise<PeriodComparison> {
  const params = new URLSearchParams({
    current_start: currentStart,
    current_end: currentEnd,
    previous_start: previousStart,
    previous_end: previousEnd,
  });

  const url = `${BASE_URL}/compare/${projectId}?${params.toString()}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to compare periods: ${response.statusText}`);
  }

  return response.json();
}

// Hook para facilitar uso com React Query
export const metricsApi = {
  getProjectMetrics,
  getTokenUsage,
  getExecutionTimes,
  getCostAnalysis,
  getTokenTrends,
  getExecutionPerformance,
  getROIMetrics,
  getProductivityMetrics,
  getInsights,
  getHourlyMetrics,
  backfillMetrics,
  comparePeriods,
};

File: backend/src/routes/metrics.py
"""Metrics routes for the API."""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional, Literal
from datetime import date, datetime, timedelta
from pydantic import BaseModel

from ..database import get_db
from ..repositories.metrics_repository import MetricsRepository
from ..services.metrics_aggregator import MetricsAggregator
from ..services.metrics_collector import MetricsCollector


router = APIRouter(prefix="/api/metrics", tags=["metrics"])


# Schemas
class MetricsResponse(BaseModel):
    """Response model for aggregated metrics."""
    totalInputTokens: int
    totalOutputTokens: int
    totalTokens: int
    totalCost: float
    avgExecutionTimeMs: int
    minExecutionTimeMs: int
    maxExecutionTimeMs: int
    totalExecutions: int
    successfulExecutions: int
    successRate: float


class TokenUsageResponse(BaseModel):
    """Response model for token usage data."""
    data: list


class ExecutionTimeResponse(BaseModel):
    """Response model for execution time data."""
    data: list


class CostAnalysisResponse(BaseModel):
    """Response model for cost analysis."""
    data: list


class InsightsResponse(BaseModel):
    """Response model for insights."""
    insights: list


class TrendsResponse(BaseModel):
    """Response model for trends."""
    movingAverage: int
    peakUsage: dict
    trend: float
    projection: int
    dailyData: list


class PerformanceResponse(BaseModel):
    """Response model for performance analysis."""
    p50: int
    p95: int
    p99: int
    mean: int
    min: int
    max: int
    stdDev: int
    outlierCount: int
    outlierThreshold: int


class ROIResponse(BaseModel):
    """Response model for ROI metrics."""
    costPerCard: float
    totalCost: float
    cardsCompleted: int
    modelEfficiency: list
    estimatedTimeSaved: dict


class ProductivityResponse(BaseModel):
    """Response model for productivity metrics."""
    cardsCompleted: int
    cardsInProgress: int
    cardsTodo: int
    velocity: float
    totalCards: int


@router.get("/project/{project_id}", response_model=MetricsResponse)
async def get_project_metrics(
    project_id: str,
    start_date: Optional[date] = Query(None),
    end_date: Optional[date] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """
    Retorna métricas agregadas do projeto.

    Args:
        project_id: ID do projeto
        start_date: Data inicial (opcional)
        end_date: Data final (opcional)
    """
    repo = MetricsRepository(db)
    metrics = await repo.get_aggregated_metrics(project_id, start_date, end_date)

    return MetricsResponse(**metrics)


@router.get("/tokens/{project_id}", response_model=TokenUsageResponse)
async def get_token_usage(
    project_id: str,
    period: Literal["24h", "7d", "30d", "all"] = Query("7d"),
    group_by: Literal["hour", "day", "model"] = Query("day"),
    db: AsyncSession = Depends(get_db)
):
    """
    Retorna uso de tokens agregado por período.

    Args:
        project_id: ID do projeto
        period: Período de análise (24h, 7d, 30d, all)
        group_by: Agrupamento (hour, day, model)
    """
    repo = MetricsRepository(db)
    data = await repo.get_token_usage(project_id, period, group_by)

    return TokenUsageResponse(data=data)


@router.get("/execution-time/{project_id}", response_model=ExecutionTimeResponse)
async def get_execution_times(
    project_id: str,
    command: Optional[str] = Query(None),
    limit: int = Query(100),
    db: AsyncSession = Depends(get_db)
):
    """
    Retorna tempos de execução por card/comando.

    Args:
        project_id: ID do projeto
        command: Filtrar por comando específico (opcional)
        limit: Limite de resultados
    """
    repo = MetricsRepository(db)
    data = await repo.get_execution_times(project_id, command, limit)

    return ExecutionTimeResponse(data=data)


@router.get("/costs/{project_id}", response_model=CostAnalysisResponse)
async def get_cost_analysis(
    project_id: str,
    group_by: Literal["model", "command", "day"] = Query("model"),
    db: AsyncSession = Depends(get_db)
):
    """
    Retorna análise de custos detalhada.

    Args:
        project_id: ID do projeto
        group_by: Agrupamento (model, command, day)
    """
    repo = MetricsRepository(db)
    data = await repo.get_cost_analysis(project_id, group_by)

    return CostAnalysisResponse(data=data)


@router.get("/trends/{project_id}", response_model=TrendsResponse)
async def get_token_trends(
    project_id: str,
    days: int = Query(7, ge=1, le=90),
    db: AsyncSession = Depends(get_db)
):
    """
    Calcula tendências de uso de tokens.

    Args:
        project_id: ID do projeto
        days: Número de dias para análise (1-90)
    """
    aggregator = MetricsAggregator(db)
    trends = await aggregator.calculate_token_trends(project_id, days)

    return TrendsResponse(**trends)


@router.get("/performance/{project_id}", response_model=PerformanceResponse)
async def get_execution_performance(
    project_id: str,
    command: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """
    Analisa performance de execuções (percentis, outliers).

    Args:
        project_id: ID do projeto
        command: Filtrar por comando específico (opcional)
    """
    aggregator = MetricsAggregator(db)
    performance = await aggregator.analyze_execution_performance(project_id, command)

    return PerformanceResponse(**performance)


@router.get("/roi/{project_id}", response_model=ROIResponse)
async def get_roi_metrics(
    project_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    Calcula métricas de ROI (custo por card, eficiência, economia de tempo).

    Args:
        project_id: ID do projeto
    """
    aggregator = MetricsAggregator(db)
    roi = await aggregator.calculate_roi_metrics(project_id)

    return ROIResponse(**roi)


@router.get("/productivity/{project_id}", response_model=ProductivityResponse)
async def get_productivity_metrics(
    project_id: str,
    start_date: Optional[date] = Query(None),
    end_date: Optional[date] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """
    Retorna métricas de produtividade (velocity, cycle time, throughput).

    Args:
        project_id: ID do projeto
        start_date: Data inicial (opcional)
        end_date: Data final (opcional)
    """
    repo = MetricsRepository(db)
    productivity = await repo.get_productivity_metrics(project_id, start_date, end_date)

    return ProductivityResponse(**productivity)


@router.get("/insights/{project_id}", response_model=InsightsResponse)
async def get_insights(
    project_id: str,
    db: AsyncSession = Depends(get_db)
):
    """
    Gera insights automáticos sobre as métricas.

    Args:
        project_id: ID do projeto
    """
    aggregator = MetricsAggregator(db)
    insights = await aggregator.generate_insights(project_id)

    return InsightsResponse(insights=insights)


@router.get("/hourly/{project_id}")
async def get_hourly_metrics(
    project_id: str,
    target_date: date = Query(default_factory=lambda: datetime.utcnow().date()),
    db: AsyncSession = Depends(get_db)
):
    """
    Retorna métricas agregadas por hora para um dia específico.

    Args:
        project_id: ID do projeto
        target_date: Data alvo (padrão: hoje)
    """
    aggregator = MetricsAggregator(db)
    hourly_data = await aggregator.aggregate_hourly_metrics(project_id, target_date)

    return {"data": hourly_data}


@router.post("/backfill/{project_id}")
async def backfill_metrics(
    project_id: str,
    start_date: Optional[datetime] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """
    Preenche métricas retroativamente para execuções existentes.

    Args:
        project_id: ID do projeto
        start_date: Data inicial para backfill (opcional)
    """
    collector = MetricsCollector(db)
    count = await collector.backfill_metrics(project_id, start_date)

    return {
        "message": f"Backfill completed successfully",
        "metricsCreated": count
    }


@router.get("/compare/{project_id}")
async def compare_periods(
    project_id: str,
    current_start: date = Query(...),
    current_end: date = Query(...),
    previous_start: date = Query(...),
    previous_end: date = Query(...),
    db: AsyncSession = Depends(get_db)
):
    """
    Compara métricas entre dois períodos.

    Args:
        project_id: ID do projeto
        current_start: Data inicial do período atual
        current_end: Data final do período atual
        previous_start: Data inicial do período anterior
        previous_end: Data final do período anterior
    """
    aggregator = MetricsAggregator(db)
    comparison = await aggregator.compare_periods(
        project_id,
        current_start,
        current_end,
        previous_start,
        previous_end
    )

    return comparison
